connectome_interpreter/utils.py

```diff
@@ -1087,3 +1087,164 @@ def path_for_ngl(path):
         df = df.drop_duplicates()
         dfs.append(df)
     return pd.concat(dfs)
+
+
+def make_grid_inputs(v1, v2, num_layers: int, grid_size: int = 10, timepoints: int | list = 0, device=None) -> Tuple[torch.Tensor, List[Tuple[float, float]]]:
+    """
+    Make a batch of input using combinations of v1 and v2 at different strengths (0 to 1). 
+
+    Args:
+        v1 (np.ndarray): The first input vector.
+        v2 (np.ndarray): The second input vector. Its length should match that of `v1`.
+        num_layers (int): The number of layers in the model.
+        grid_size (int, optional): The number of points in the grid. Defaults to 10.
+        timepoints (int|list, optional): The timepoints at which combinations of v1 and v2 are used. Defaults to 0 (the first timepoint).
+        device (str, optional): The device to create the inputs on. If None, if GPU is available, the inputs are created on the GPU. Otherwise CPU.
+
+    Returns:
+        torch.Tensor: A tensor of shape (grid_size**2, len(v1), num_layers).
+        list: A list of grid coordinates.
+    """
+
+    # first check if length of v1 and v2 are the same
+    if len(v1) != len(v2):
+        raise ValueError("The length of v1 and v2 should be the same.")
+
+    if device is None:
+        device = torch.device("cuda" if torch.cuda.is_available() else "cpu")
+
+    sc1 = np.linspace(0, 1, grid_size)
+    sc2 = np.linspace(0, 1, grid_size)
+    sc1_grid, sc2_grid = np.meshgrid(sc1, sc2)
+
+    inputs = np.zeros((grid_size**2, len(v1), num_layers), dtype=np.float32)
+    for i, (s1, s2) in enumerate(zip(sc1_grid.ravel(), sc2_grid.ravel())):
+        inputs[i, :, timepoints] = v1 * s1 + v2 * s2
+
+    # Convert to torch tensor
+    inputs_tensor = torch.from_numpy(inputs).to(device)
+
+    # Create grid coordinates for reference
+    grid_coords = list(zip(sc1_grid.ravel(), sc2_grid.ravel()))
+
+    return inputs_tensor, grid_coords
+
+
+def plot_output_grid(grid_outputs: torch.Tensor, grid_coords: List[Tuple[float, float]], selected_index: int | List[int], figsize: tuple = (10, 8)):
+    """
+    Plot the output grid for selected indices.
+
+    Args:
+        grid_outputs (torch.Tensor): The output grid tensor of shape (grid_size**2, num_neurons, num_layers).
+        grid_coords (list): A list of grid coordinates. Best from function `make_grid_inputs()`.
+        selected_index (int|list): The index or indices to plot.
+        figsize (tuple, optional): The size of the figure. Defaults to (10, 8).
+    """
+    selected_index = to_nparray(selected_index)
+
+    def plot_heatmap(index):
+        # Create the plot
+        plt.figure(figsize=figsize)
+        im = plt.imshow(heatmaps[index-1], cmap='viridis', origin='lower',
+                        extent=[0, 1, 0, 1])
+        plt.colorbar(im, label='Output')
+        plt.xlabel('v1 Activation')
+        plt.ylabel('v2 Activation')
+        plt.title(f'Output Grid for {selected_index}')
+        plt.show()
```

Comment:
I tend to prefer using the object-oriented model
```suggestion
    def plot_heatmap(index: int) -> plt.Figure:
        # Create the plot
        fig,  ax = plt.subplots(1, 1, figsize=figsize)
        im = aximshow(heatmaps[index-1], cmap='viridis', origin='lower',
                        extent=[0, 1, 0, 1])
        ax.colorbar(im, label='Output')
        ax.xlabel('v1 Activation')
        ax.ylabel('v2 Activation')
        ax.title(f'Output Grid for {selected_index}')
        fig.show()  # see below... maybe you don't want it here
        return fig  # see below
```